includes/create-theme/theme-token-processor.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * Fine grained token processing class.
+ */
+class CBT_Token_Processor {
+	private $p;
+	private $tokens           = array();
+	private $text             = '';
+	private $translators_note = '/* Translators: ';
+	private $increment        = 0;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string $string The string to process.
+	 */
+	public function __construct( $string ) {
+		$this->p = new WP_HTML_Tag_Processor( $string );
+	}
+
+	/**
+	 * Processes the HTML tags in the string and updates tokens, text, and translators' note.
+	 *
+	 * @param $p The string to process.
+	 * @return void
+	 */
+	public function process_tokens() {
+		while ( $this->p->next_token() ) {
+			$token_type    = $this->p->get_token_type();
+			$token_name    = strtolower( $this->p->get_token_name() );
+			$is_tag_closer = $this->p->is_tag_closer();
+
+			if ( '#tag' === $token_type ) {
+				$this->increment++;
+				$this->text .= '%' . $this->increment . '$s';
+
+				if ( 1 !== $this->increment ) {
+					$this->translators_note .= ', ';
+				}
+
+				if ( $is_tag_closer ) {
+					$this->tokens[]          = "</{$token_name}>";
+					$this->translators_note .= '%' . $this->increment . "\$s is the end of a '" . $token_name . "' HTML element";
+				} else {
+					$token      = '<' . $token_name;
+					$attributes = $this->p->get_attribute_names_with_prefix( '' );
+
+					foreach ( $attributes as $attr_name ) {
+						$attr_value = $this->p->get_attribute( $attr_name );
+						$token     .= $this->process_attribute( $attr_name, $attr_value );
+					}
+
+					$token                  .= '>';
+					$this->tokens[]          = $token;
+					$this->translators_note .= '%' . $this->increment . "\$s is the start of a '" . $token_name . "' HTML element";
```

Comment:
Do we have any way of differentiating self closing tags? For example "%3$s is the start of a 'br' HTML element" might be a little confusing, since it doesn't (usually) have a closing tag.

Reply:
Handled in https://github.com/WordPress/create-block-theme/pull/719/commits/6b5fe21f2372f6918cb0e1f6e825b7721d44cfab.